docs/tsm/TSI_CARDINALITY_PROPOSAL.md

```diff
@@ -0,0 +1,254 @@
+
+##### Building an index sketch
+
+When a new index is created as the result of a full compaction, a new sketch will be generated for both the series and measurements within the index.
+This will involve scanning and hashing with a 64-bit hash function, every series and measurement value within the index.
+A suitable hash function for this purpose would be [MurmurHash3][4], which is optimised to generate 128-bit hash values on 64-bit architectures (we can use the first 64 bits for the series/measurement hash).
+Each hashed value will then contribute to the HLL++ sketch for either the series or measurement sketch.
+
+Some prototyping and testing may be needed, but I would suggest a starting point for the precision of the HLL++ sketch to be `14`, which will require a sketch size of around `16KB`.
+
+Once the sketch has been generated it can be stored in the index and also in memory within the `Shard`.
+
+##### Getting cardinality estimates
+
+Cardinality estimates can be provided in constant time by consulting a series or measurement sketch.
+If series or measurements are distributed across multiple indices (likely), then those sketches can be combined in linear time, with respect to the size of the sketch, which typically will be small (`~4096`).
+In summary, cardinality estimations will be fast, and will not be affected by the number of series or measurements in a database.
+
+When calculating estimates, we will need to consider the number of tombstoned series/measurements for the database, and subtract them from the cardinality estimate.
+
+##### Adding new series and measurements
+
+Since TSI indices are immutable, we should not modify an index's sketch when a new series is added to the database.
+Instead we will update the in-memory sketches, and for durability we will then also update sketches we will keep in the WAL files.
+
+##### Removing series and measurements
+
+The HyperLogLog family of algorithms unfortunately do not support removing values from the sketch, since the sketch itself is lossy and does not contain enough information to support deletions.
+However, we can improve the accuracy of estimations by keeping track of series and measurements that are deleted from the database.
+
+In the case of series removal, a tombstone entry is added to the WAL.
+When this occurs we will increment a series tombstone counter for that database.
+As discussed above, this count will be subtracted from any cardinality estimates for the database.
+
+In the case of measurement removal, we will maintain a similar counter for removed measurements.
+However, when a measurement is removed, all series for that measurement are removed, so we need to get a count of the number of series belonging to the measurement, in order to increase the removed series counter too.
+We can do this by consulting the measurement hash index in the TSI
+and looking up the `len(series)` value, which tells us how many series belong to the measurement.
```

Comment:
The exact number of series is available in the TSI file. See the [measurements block here](https://github.com/benbjohnson/influxdb/blob/942c552666a119d6866f22bc1d0c70157ee4dbf3/doc/tsm/tsi.md#measurements-block).